### PR TITLE
Enum type generator

### PR DESCRIPTION
**Pull request by jiahuif**
This PR is the implementation of https://github.com/kubernetes/enhancements/issues/2887 in k/kube-openapi
This PR implements generation of enum type definition in the schema.

- Detect enum types by `+enum`
- Adds all possible values to [Enum](https://github.com/kubernetes/kube-openapi/blob/95288971da7e1e3d736b57c826a238a14330e0a8/pkg/validation/spec/schema.go#L180) field
- Add a description of all enum values, as a Markdown list, to the description.

An example:
```go
package foo

// EnumType is the enumType
// +enum
type EnumType string

// EnumA is a
const EnumA EnumType = "a"
// EnumB is b
const EnumB EnumType = "b"

// Blah is a test.
// +k8s:openapi-gen=true
// +k8s:openapi-gen=x-kubernetes-type-tag:type_test
type Blah struct {
  // Value is the value
	Value EnumType
}
```

generated:
```go
common.OpenAPIDefinition{
	Schema: spec.Schema{
		SchemaProps: spec.SchemaProps{
			Description: "Blah is a test.",
			Type:        []string{"object"},
			Properties: map[string]spec.Schema{
				"Value": {
					SchemaProps: spec.SchemaProps{
						Description: "Value is the value EnumType:\n - `a`: is a\n - `b`: is b",
						Default:     "",
						Type:        []string{"string"},
						Format:      "",
						Enum:        []interface{}{"a", "b"}},
				},
			},
			Required: []string{"Value"},
		},
		VendorExtensible: spec.VendorExtensible{
			Extensions: spec.Extensions{
				"x-kubernetes-type-tag": "type_test",
			},
		},
	},
}

```

### CODE REVIEW

**Comment by jiahuif** [comment]
@apelisse Is it okay to bump `k8s.io/gengo` version? The current version is very old and causing issues with my work.

**Comment by apelisse** [comment]
Sure!

**Comment by jiahuif** [comment]
Also
/cc @jpbetz

File: test/integration/pkg/generated/openapi_generated.go [hunk 1]
```diff
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
```
Comment (jiahuif) [inline test/integration/pkg/generated/openapi_generated.go:1]: Note: this is added through `go tidy`

File: pkg/generators/enum.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+		v = append(v, fmt.Sprintf("%q", value.Value))
+	}
+	sort.Strings(v)
+	return
+}
+
+// FormatEnumDescription returns a description of the enum type in the following
+//  format:
+// FooType:
+//  - `value1`: description 1
```
Comment (jpbetz) [inline pkg/generators/enum.go:72]: very minor nit: The below code writes the bullet list with `*`, not `-`

File: pkg/generators/enum.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	sort.Strings(lines)
+	return append([]string{header}, lines...)
+}
+
+func parseEnums(c *generator.Context) (enumTypes enumMap) {
+	stringType := c.Universe.Type(types.Name{Name: "string"})
+
+	enumTypes = make(enumMap)
+	for _, p := range c.Universe {
+
```
Comment (jpbetz) [inline pkg/generators/enum.go:101]: nit: empty line needed here?

File: pkg/generators/enum.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	"k8s.io/gengo/generator"
+	"k8s.io/gengo/types"
+)
+
+const tagEnumType = "enum"
+
+type enumValue struct {
+	Name    string
+	Value   string
+	Comment string
```
Comment (jpbetz) [inline pkg/generators/enum.go:34]: An alternative to keeping track of the comments of each enum value would be to keep track of the comment for the entire enum? E.g.:

`go
type enumType struct {
  Values []*enumValue
  Description string
}

type enumValue struct {
  Name string
  Value string
}
`

The advantage being that the description only needs to be constructed once.

File: pkg/generators/openapi_test.go [hunk 5]
```diff
@@ -1570,3 +1570,60 @@ map[string]interface{}{
 
 `, funcBuffer.String())
 }
+
+func TestEnum(t *testing.T) {
+	callErr, funcErr, assert, _, funcBuffer := testOpenAPITypeWriter(t, `
+package foo
+
+// EnumType is the enumType
```
Comment (jpbetz) [inline pkg/generators/openapi_test.go:1578]: I was expecting this enum type comment to be included along with the enum value comments.

**Review by jpbetz — CHANGES_REQUESTED** [review]

File: pkg/generators/enum.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+// IsKnownEnumType checks if a given type is a known enum type
+func (ec *enumContext) IsKnownEnumType(t *types.Type) bool {
+	_, ok := ec.enumTypes[t.Name]
+	return ok
+}
+
+// FormatEnumValues returns all possible values of the enum type
+//  the results are sorted and quoted as Go literals.
```
Comment (apelisse) [inline pkg/generators/enum.go:55]: Nit: A lot of your comments have a weird structure/indentation/spaces inserted.

File: pkg/generators/enum.go (hunk 2 above)
Comment (apelisse) [reply pkg/generators/enum.go:72]: By the way, I think I like dash more than star, but I think that's completely subjective so mostly ignore me ;-)

File: pkg/generators/enum.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+//  - `value1`: description 1
+//  - `value2`: description 2
+func (ec *enumContext) FormatEnumDescription(t *types.Type) []string {
+	values, ok := ec.enumTypes[t.Name]
+	if !ok {
+		return nil
+	}
+
+	var lines []string
+	header := t.Name.Name + ":"
```
Comment (apelisse) [inline pkg/generators/enum.go:81]: I don't know about the Go type name, since that's an implementation detail

File: pkg/generators/enum.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	for _, value := range values {
+		comment := strings.TrimSpace(value.Comment)
+
+		// The comment should starts with the type name, trim it first.
+		comment = strings.TrimPrefix(comment, value.Name)
+		comment = strings.TrimSpace(comment)
+		// The comment may be multiline, cascade all consecutive whitespaces.
+		comment = whitespaceRegex.ReplaceAllString(comment, " ")
```
Comment (apelisse) [inline pkg/generators/enum.go:89]: Maybe isolate all of this in a separate function, I found it a little confusing as I was reading the code.

File: pkg/generators/enum.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+		lines = append(lines, fmt.Sprintf(" * `%s`: %s", value.Value, comment))
+	}
+	sort.Strings(lines)
+	return append([]string{header}, lines...)
+}
+
+func parseEnums(c *generator.Context) (enumTypes enumMap) {
```
Comment (apelisse) [inline pkg/generators/enum.go:96]: I'm not generally a big fan of named returned values, but that's probably just my preference.

**Review by apelisse — COMMENTED** [review]
Looks good overall. I can see now why you're asking about generating v3 OpenAPI, let's talk about that.

File: test/integration/pkg/generated/openapi_generated.go [hunk 10]
```diff
@@ -276,6 +278,34 @@ func schema_test_integration_testdata_dummytype_Waldo(ref common.ReferenceCallba
...
+			SchemaProps: spec.SchemaProps{
+				Type: []string{"object"},
+				Properties: map[string]spec.Schema{
+					"content": {
+						SchemaProps: spec.SchemaProps{
+							Description: "\n\nFruitType:\n * `apple`: is the Apple\n * `banana`: is the Banana\n * `onigiri`: is the Rice ball that does not seem to belong to a fruits basket but has a long comment that spans multiple lines",
+							Default:     "",
+							Type:        []string{"string"},
+							Format:      "",
+							Enum:        []interface{}{"apple", "banana", "onigiri"}},
```
[the remark this answers is not in the document]
Comment (apelisse) [reply test/integration/pkg/generated/openapi_generated.go:293]: Ah, if they exist in v2 (which they seem to do: https://swagger.io/docs/specification/2-0/enums/), then we should totally add them right away and distinguish that effort from v3? Or am I missing something?

[the remark this answers is not in the document]
Comment (jiahuif) [reply test/integration/pkg/generated/openapi_generated.go:293]: As long as nobody is relying on the enum fields being empty.

[the remark this answers is not in the document]
Comment (jiahuif) [reply test/integration/pkg/generated/openapi_generated.go:293]: Reusable enums is another improvement of v3 over v2. If we are to use this feature, the generation part would need checking the version anyways.

Ref:  https://swagger.io/docs/specification/data-models/enums/

File: pkg/generators/enum.go (hunk 6 above)
Comment (jiahuif) [reply pkg/generators/enum.go]: Misconfiguration of the editor (with my own workstation). Fixed.

File: pkg/generators/enum.go (hunk 7 above)
Comment (jiahuif) [reply pkg/generators/enum.go:81]: Good point. I am rethinking how this header should be structured.

An example from https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#containerport-v1-core
Field | Description
-- | --
protocol string Protocol for port. Must be UDP, TCP, or SCTP. Defaults to "TCP".

File: test/integration/pkg/generated/openapi_generated.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (apelisse) [reply test/integration/pkg/generated/openapi_generated.go:293]: > As long as nobody is relying on the enum fields being empty.

Yeah, I'm not concerned by that.

> Reusable enums is another improvement of v3 over v2. If we are to use this feature, the generation part would need checking the version anyways.

Yeah, we could consider these, but we don't have too at that time. I think we're fine just duplicating things wherever needed for now.

File: pkg/generators/enum.go (hunk 3 above)
Comment (jiahuif) [reply pkg/generators/enum.go:94]: was a removed comment, I will polish it a bit.

File: pkg/generators/enum.go (hunk 4 above)
Comment (jiahuif) [reply pkg/generators/enum.go:34]: The comment of entire enum will be used for at most once. I think a getter would be better

File: pkg/generators/enum.go (hunk 7 above)
Comment (jiahuif) [reply pkg/generators/enum.go]: I replace the header with a simple `Possible enum values:`

File: pkg/generators/enum.go (hunk 9 above)
Comment (jiahuif) [reply pkg/generators/enum.go]: There are cases where multiple return values can be more easily identified with named returned values. I left these cases but removed the rest.

File: pkg/generators/openapi_test.go (hunk 5 above)
Comment (jiahuif) [reply pkg/generators/openapi_test.go]: The comment on the enum type, like the name of the enum type, is usually kind of implementation detail. For example.

```go
// Protocol defines network protocols supported for things like container ports.
type Protocol string
```

later
```go
// ServicePort represents the port on which the service is exposed
type ServicePort struct {
	// Optional if only one ServicePort is defined on this service: The
	// name of this port within the service.  This must be a DNS_LABEL.
	// All ports within a ServiceSpec must have unique names.  This maps to
	// the 'Name' field in EndpointPort objects.
	Name string

	// The IP protocol for this port.  Supports "TCP", "UDP", and "SCTP".
	Protocol Protocol
```
The comment on the field should be good enough.

File: pkg/generators/enum.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+		// use "%q" format to generate a Go literal of the string const value
+		values = append(values, fmt.Sprintf("%q", value.Value))
+	}
+	sort.Strings(values)
+	return values
+}
+
+// DescriptionLines returns a description of the enum in this format:
+//
+// FooType:
```
Comment (apelisse) [inline pkg/generators/enum.go:76]: extremely minor nit, but that comment is out of date now

File: pkg/generators/enum.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+//  -	`value1`: description 1
+//  - `value2`: description 2
+func (et *enumType) DescriptionLines() []string {
+	var lines []string
+	for _, value := range et.Values {
+		lines = append(lines, value.Description())
```
Comment (apelisse) [inline pkg/generators/enum.go:82]: Thanks, this function is much easier to read!

**Review by apelisse — COMMENTED** [review]
Beside a minor comment nit,  I don't have anything against that. We will need an approved KEP to merge it I think.

File: pkg/generators/openapi_test.go (hunk 5 above)
Comment (jpbetz) [reply pkg/generators/openapi_test.go]: I agree that this is often not a particularly valuable comment.

But if a developer does add some really important information in this and we choose to omit it, and then as a result the API is misused, I don't think I could defend our decision to omit the the comment. I think we responsible for including everything author included in the godoc of the type in the openapi documentation.

**Comment by jiahuif** [comment]
> Beside a minor comment nit, I don't have anything left to say. We will need an approved KEP to merge it I think.

KEP is there. Let's restart this.

File: pkg/generators/enum.go (hunk 11 above)
Comment (jiahuif) [reply pkg/generators/enum.go:76]: Good catch, thanks!

File: test/integration/pkg/generated/openapi_generated.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (jiahuif) [reply test/integration/pkg/generated/openapi_generated.go:292]: This discussion has been done in the KEP. Resolving here.

File: pkg/generators/enum.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+	Value   string
+	Comment string
+}
+
+// enumMap is a map of enum type name to all its values
+type enumMap map[types.Name][]*enumValue
+
+type enumContext struct {
+	enumTypes enumMap
+}
```
[the remark this answers is not in the document]
Comment (jiahuif) [reply pkg/generators/enum.go]: actually, reworked this part.

File: pkg/generators/enum.go (hunk 2 above)
Comment (jiahuif) [reply pkg/generators/enum.go:77]: I did a quick count and find out `*` is rarely used in our code base. So I switched to dash.

**Comment by Jefftree** [comment]
@jiahuif it might be useful to have an additional PR to show how the k/k native OpenAPI changes as a result of this.

**Comment by apelisse** [comment]
> @jiahuif it might be useful to have an additional PR to show how the k/k native OpenAPI changes as a result of this.

Totally, I'd love to see a WIP pull-request in kubernetes that integrates this change locally and adds the enum marker in a few places to see what it does.

**Comment by jiahuif** [comment]
@Jefftree @apelisse PoC sent as https://github.com/kubernetes/kubernetes/pull/105018

File: test/integration/testdata/enumtype/enum.go [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+// FruitBanana is the Banana
+const FruitBanana FruitType = "banana"
+
+// FruitRiceBall is the Rice ball that does not seem to belong to
+// a fruits basket but has a long comment that is so long that it spans
+// multiple lines
+const FruitRiceBall FruitType = "onigiri"
+
+// FruitsBasket is the type that contains the emun type.
```
Comment (Jefftree) [inline test/integration/testdata/enumtype/enum.go:18]: s/emun/enum

**Review by Jefftree — COMMENTED** [review]
Thanks, this looks great! One nit, otherwise lgtm

File: test/integration/testdata/enumtype/enum.go (hunk 14 above)
Comment (jiahuif) [reply test/integration/testdata/enumtype/enum.go]: fixed. Thanks!

**Pull request merged**